tikv/compatible_txn_safe_point_loader.go

```diff
@@ -0,0 +1,136 @@
+package tikv
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/tikv/client-go/v2/internal/apicodec"
+	"github.com/tikv/client-go/v2/util"
+	pd "github.com/tikv/pd/client"
+	etcdmvccpb "go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+const (
+	unifiedTxnSafePointPath       = "/tidb/store/gcworker/saved_safe_point"
+	keyspaceLevelTxnSafePointPath = "/keyspaces/tidb/%d/tidb/store/gcworker/saved_safe_point"
+)
+
+// compatibleTxnSafePointLoader is used to load txn safe point from etcd for old versions where the GetGCState API
+// is not yet supported.
+//
+// The same functionality can be provided by EtcdSafePointKV. However, EtcdSafePointKV determines whether the key needs
+// to be prefixed. Meanwhile, when client-go's user creates KVStore, it needs manually create EtcdSafePointKV first.
+// This means, whether the etcd client is correctly prefixed depends on whether the user creates it correctly.
+// Considering that TiDB is not the only user of client-go, this was a bad design, and we cannot assume the caller
+// can easily make it correct (e.g., whether the txn safe point is stored in prefixed key depends on whether
+// keyspace level GC is enabled for a keyspace). To minimize the risk of misusing, we rewrote a special util for loading
+// it. The EtcdSafePointKV, on the contrary, will be deprecated in the future.
+type compatibleTxnSafePointLoader struct {
+	mu        sync.Mutex
+	etcdCli   atomic.Pointer[clientv3.Client]
+	endpoints []string
+	tlsConfig *tls.Config
+	codec     apicodec.Codec
+}
+
+func newCompatibleTxnSafePointLoader(codec apicodec.Codec, etcdEndpoints []string, tlsConfig *tls.Config) *compatibleTxnSafePointLoader {
+	return &compatibleTxnSafePointLoader{
+		endpoints: etcdEndpoints,
+		tlsConfig: tlsConfig,
+		codec:     codec,
+	}
+}
+
+func (l *compatibleTxnSafePointLoader) getEtcdCli() (*clientv3.Client, error) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	cli := l.etcdCli.Load()
```

Comment:
This is under the protect of mutex, we can just use `cli := l.etcdCli` and `l.etcdCli = cli` ?

Reply:
The first attempt to load it is in `loadTxnSafePoint` and is not protected by the mutex. In this case read outside the mutex can be parallel with the write in the mutex. It's still unsafe without memory barrier.